server/remote_cache/digest/digest.go

```diff
@@ -454,95 +438,148 @@ func ComputeForFile(path string, digestType repb.DigestFunction_Value) (*repb.Di
+	rn, err := parseResourceName(resourceName, rspb.CacheType_CAS, uploadResourceName)
 	if err != nil {
 		return nil, err
 	}
 	return rn.CheckCAS()
 }
 
 func ParseDownloadResourceName(resourceName string) (*CASResourceName, error) {
-	rn, err := parseResourceName(resourceName, downloadRegex, rspb.CacheType_CAS)
+	rn, err := parseResourceName(resourceName, rspb.CacheType_CAS, downloadResourceName)
 	if err != nil {
 		return nil, err
 	}
 	return rn.CheckCAS()
 }
 
 func ParseActionCacheResourceName(resourceName string) (*ACResourceName, error) {
-	rn, err := parseResourceName(resourceName, actionCacheRegex, rspb.CacheType_AC)
+	rn, err := parseResourceName(resourceName, rspb.CacheType_AC, downloadResourceName)
 	if err != nil {
 		return nil, err
 	}
 	return rn.CheckAC()
 }
 
 func IsDownloadResourceName(url string) bool {
-	return isResourceName(url, downloadRegex)
+	_, err := parseResourceName(url, rspb.CacheType_CAS, downloadResourceName)
+	return err == nil
 }
 
 func IsActionCacheResourceName(url string) bool {
-	return isResourceName(url, actionCacheRegex)
+	_, err := parseResourceName(url, rspb.CacheType_AC, downloadResourceName)
+	return err == nil
+}
+
+func parseResourceName(resourceName string, cacheType rspb.CacheType, resourceType resourceNameType) (*ResourceName, error) {
+
+	pieces := strings.Split(resourceName, "/")
+
+	// Need at least 2 slashes for CAS: blobs/hash/size
+	// 3 for AC: blobs/ac/hash/size
+	// 4 for upload: uploads/uuid/blobs/hash/size
+	if len(pieces) < 3 || (cacheType == rspb.CacheType_AC && len(pieces) < 4) || (resourceType == uploadResourceName && len(pieces) < 5) {
+		return nil, status.InvalidArgumentErrorf("Unparseable resource name, not enough pieces: %s", resourceName)
+	}
+
+	// Parse back-to-front to avoid having to find the end of the instance name
+	// The last piece must be the size (in bytes)
+	pieceIdx := len(pieces) - 1
+	sizeBytes, err := strconv.ParseInt(pieces[pieceIdx], 10, 64)
+	if err != nil {
+		return nil, status.InvalidArgumentErrorf("Unparseable resource name, invalid size: %s", resourceName)
+	}
+	if sizeBytes < 0 {
+		return nil, status.InvalidArgumentErrorf("Unparseable resource name, negative size: %s", resourceName)
+	}
+
+	// The second-to-last piece is the hash
+	pieceIdx--
+	hash := pieces[pieceIdx]
+	d := &repb.Digest{Hash: hash, SizeBytes: sizeBytes}
+	inferredDigestFunction := InferOldStyleDigestFunctionInDesperation(d)
+	digestFunction := inferredDigestFunction
+	if inferredDigestFunction == repb.DigestFunction_UNKNOWN {
+		return nil, status.InvalidArgumentErrorf("Unparseable resource name, invalid hash (wrong length): %s", resourceName)
+	}
+	if !isLowerHex(hash) {
+		return nil, status.InvalidArgumentErrorf("Unparseable resource name, invalid hash (bad character(s)): %s", resourceName)
+	}
+
+	// The next piece may be the digest function
+	pieceIdx--
+	piece := pieces[pieceIdx]
+	if piece == "blake3" {
```

Comment:
is it valid (according to the spec) to set other digest functions here?

Reply:
It is. Added the other functions.